### DISCUSSION TITLE
2.0.0

### DISCUSSION THREAD

**Discussion by cpisciotta** [cpisciotta/xcbeautify#268] [Announcements]
### Types

- `Parser` and `OutputHandler` are now internal and replaced by [XCBeautifier](https://github.com/cpisciotta/xcbeautify/blob/main/Sources/XcbeautifyLib/XCBeautifier.swift), the single entry point to parse xcodebuild commands. You can find an example migration [here](https://github.com/tuist/tuist/pull/5886).
- `OutputType` and `JunitReporter` are now internal. JUnit reporting is now only available via the command-line executable.

### Configuration

- The project now uses Swift Tools Version 5.9, and any dependent projects must update their minimum version to at least this version.
- The project no longer supports local development using Bazel.

## Note

With these changes, it might be more difficult for dependent projects to modify the formatted values. If you'd like to specify a custom formatting strategy, such as for a new CI provider, it is likely best to create a new `OutputRendering` type. You can find the available ones [here](https://github.com/cpisciotta/xcbeautify/tree/main/Sources/XcbeautifyLib/Renderers). If you'd like to contribute one, feel free to reach out, open a pull request, or open a GitHub issue!

## What's Changed

* Introduce ParsingTests by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/235
* Delete OutputType+CustomStringConvertible by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/239
* Enable Dependabot by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/240
* Treat Warnings as Errors by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/241
* Drop Bazel Support by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/182
* Delete Unnecessary Homebrew Formula by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/245
* Clean Up Release Workflow by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/246
* Bump Minimum Swift Version to 5.9 by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/218
* Bump codecov/codecov-action from 3 to 4 by @dependabot in https://github.com/cpisciotta/xcbeautify/pull/247
* Use Package Access Modifier by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/249
* Introduce XCBeautifier by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/250
* Add SwiftCompile Support by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/254
* Delete .ruby-version by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/260
* Bump Minimum SwiftFormat to 0.53.1 by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/261
* Rename assets to .readme-images by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/262
* Delete XCTestManifests.swift by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/263
* Add measure script by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/264
* Implement WriteAuxiliaryFileCaptureGroup by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/257
* Add CompilationResultCaptureGroup by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/258
* Add SwiftDriverJobDiscoveryEmittingModuleCaptureGroup by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/259
* Update README by @cpisciotta in https://github.com/cpisciotta/xcbeautify/pull/266

**Full Changelog**: https://github.com/cpisciotta/xcbeautify/compare/1.4.0...2.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/cpisciotta/xcbeautify/releases/tag/2.0.0'>2.0.0</a>.</em>

**Comment by finestructure** [comment]
> Description with breaking changes and migration guide coming soon!

I was just looking into upgrading from 1.x to 2.4.1 and I'm hitting some source incompatibilities around `Parser` and `OutputHandler`. Before I dig in deeper I was wondering if the migration guide is perhaps just around the corner - or if I've not looked in the right paces! :)

**Reply by cpisciotta** [reply]
Hi @finestructure. Thanks for reaching out, and sorry for the delayed response. I've updated the description above.

In your case, I think the answer is to replace the two with `XCBeautifier`. It's worth noting that `XCBeautifier` doesn't currently support the filtering logic that `OutputHandler` does; `OutputHandler` and `OutputType` are now internal types, and `XCBeautifier` doesn't know about either. I think it makes sense to add the logic, but I haven't yet spent the time to make it happen.

Let me know if you're still encountering some issues... If so, feel free to provide more context, and I'd be happy to take a look at your use case.

**Reply by finestructure** [reply]
Thanks for the quick reply, @cpisciotta! I'll take a look and will try to isolate the use cases (as that part of the project is closed source).

**Reply by finestructure** [reply]
Ok, I think I'm _almost_ there, except for a replacement for `outputHandler.write`. Here's the original 1.x code I'm trying to upgrade:

```swift
struct Xcbeautify {
    var parser: Parser
    var outputHandler: OutputHandler

    func print(_ line: String, prefix: String = "") -> Void {
        guard let line = parser.parse(line: line) else { return }
        outputHandler.write(parser.outputType, prefix + line)
    }
}


extension Xcbeautify {
    static let console = Self.init(
        parser: .init(
            colored: false,
            renderer: .terminal,
            preserveUnbeautifiedLines: true,
            additionalLines: { nil }
        ),
        outputHandler: .init(
            quiet: false,
            quieter: false,
            isCI: true) {
                Swift.print($0)
            }
    )
}
```

`Parser` → `XCBeautifier` is straightforward but it's not clear to me how to output the parsed line like I did before:

```swift
struct Xcbeautify {
    var beautifier: XCBeautifier
    var outputHandler: OutputHandler

    func print(_ line: String, prefix: String = "") -> Void {
        guard let line = beautifier.format(line: line) else { return }
        // How can I output the parsed line (+ its optional prefix)?
        outputHandler.write(parser.outputType, prefix + line)
    }
}


extension Xcbeautify {
    static let console = Self.init(
        beautifier: XCBeautifier(
            colored: false,
            renderer: .terminal,
            preserveUnbeautifiedLines: true,
            additionalLines: { nil }
        ),
        outputHandler: .init(
            quiet: false,
            quieter: false,
            isCI: true) {
                Swift.print($0)
            }
    )
}
```

**Reply by finestructure** [reply]
Mm, from looking at how `OutputHandler` is set up in the executable target
```
        let output = OutputHandler(quiet: quiet, quieter: quieter, isCI: isCi) { print($0) }
```
I suppose this essentially just routes through `print` and I can use that instead, can't I?

**Reply by cpisciotta** [reply]
@finestructure Yes, that's correct. You can use `print` directly to replace `OutputHandler`. `OutputHandler` always uses `print`, except when `quiet` or `quieter` is applied. When one of them is applied, `OutputHandler` conditionally filters output before `print`.
